### ISSUE TITLE
Provide mechanism for conveying progress information for long running tasks

### ISSUE THREAD

**Issue by jreznot** [gradle/gradle#3654]
We have a couple of heavy network related tasks that produce progress bar in console output, like this:

```
Uploaded 10%
Uploaded 20%
...
```

At the moment, we use internal Gradle ProgressLogger class:
```
@Inject
protected ProgressLoggerFactory getProgressLoggerFactory() {
    throw new UnsupportedOperationException()
}

def progressLogger = getProgressLoggerFactory().newOperation(getClass())
progressLogger.start("Build UberJAR", getName())

progressLogger.progress("10% done);
```

It enables us to show progress in single line:
```
> :task Uploaded 10%
```

There's no really convenient way to do this (using public API).

Previously, there was the issue in JIRA: https://issues.gradle.org/browse/GRADLE-2787

It is still required for heavy long running tasks.

### Expected Behavior

It would be great if we can simply use project.logger.progress("") from tasks to show the desired single line progress output.
```
> :task some progress 10%
```

### Current Behavior

Currently, we could use either internal ProgressLogger our print all the progress to the console with multiple lines.

### Context

We want to show single line progress in percents for developers during long running tasks.

**Comment by rekire** [comment]
This would be helpful for the new `WorkerExecutor`s where the `WorkAction`s might take longer.

I want to provide the current work step there are basically 4 steps which I want to show as a progress indicator. It is something like an exporter and for my use cases it is likely that there are e.g. 4 generated tasks (based on the configuration) and are executed in parallel to safe some time (it is about an external server).

With simple logging you cannot see to which actual task the update belongs to.

**Comment by shoaniki** [comment]
This is still missing AFAICT in Gradle 7.0, and would still be extremely useful for providing users with feedback from long-running tasks such as downloads.  I cannot see any viable alternative at present to the internal API -- this isn't something we can implement for ourselves since Gradle owns the console output, and line-based logging leads to a very poor user experience, particularly if tasks are being executed in parallel.

In short, please consider providing a public API that allows a WorkAction to report its progress!

**Comment by eskatos** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team but this area of Gradle is currently not a focus one so it might take a while before it gets implemented.

**Comment by TWiStErRob** [comment]
Could this get a bit of a priority? The community has been asking for it since Gradle 1.6 in 2013 :)
The functionality is there as far as I can see: all the unit test progress, and compiling gradle.kts, storing/loading CC is the display we would like to achieve with a public API.

There are [hundreds of public usages of the "workaround"](https://github.com/search?q=org.gradle.internal.logging.progress.ProgressLoggerFactory+-path%3Asubprojects%2F+-path%3Aplatforms%2F&type=code) I would guess that each of those should count as an upvote on this issue 😉.
